### ISSUE TITLE
[java] ImmutableField false positive with inner class

### ISSUE THREAD

**Issue by kevin-wayne** [pmd/pmd#2075]
**Affects PMD Version:** 
6.15

**Rule:**
[ImmutableField](https://pmd.github.io/latest/pmd_rules_java_design.html#immutablefield)

**Description:**
ImmutableField produces a false positive when inner class modifies an instance variable in outer class.

**Code Sample demonstrating the issue:**

```java
public class Stack {
    private Node first;
    private int n;   // can't be made final

    public void insert(int x) {
        first = new Node(first, x);
    }

    public int size() {
        return n;
    }

    private class Node {
        Node next;
        int x;
        private Node(Node next, int x) {
            this.next = next;
            this.x = x;
            n++;  // inner class updates instance variable in outer class
        }
    }
}
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
CLI.
```
% pmd-bin-6.15.0/bin/run.sh pmd -no-cache -rulesets rules.xml -dir Stack.java
Stack.java:3:	Private field 'n' could be made final; it is only initialized in the declaration or constructor.
```

Using an inner class to update an instance variable in the outer class may be dubious design, but the field 'n' cannot be made final.

**Comment by kevin-wayne** [comment]
Thanks for the fix!
